pkg/resource/container_image.go

```diff
@@ -0,0 +1,469 @@
+			Size:   layer.Size,
+		}
+		layerBlob, _, err := imageSource.GetBlob(ctx, blobInfo, cache)
+		if err != nil {
+			return fmt.Errorf("could not download layer blob: %v", err)
+		}
+
+		err = r.extractLayer(layerBlob, layer, path)
+		if err != nil {
+			return fmt.Errorf("could not decompress layer: %v", err)
+		}
+		err = layerBlob.Close()
+		if err != nil {
+			return fmt.Errorf("could not close layer: %v", err)
+		}
+	}
+
+	return nil
+}
+
+func (r *ContainerImage) writeFile(filename string, content []byte) error {
+	return os.WriteFile(filepath.Join(r.clonePath, filename), content, 0600)
+}
+
+// The decompression process is a little more involved so separated out.
+func (r *ContainerImage) extractLayer(t io.Reader, layer manifest.LayerInfo, path string) error {
+	// The maximum file size should be less than 10x the layer size.
+	size := layer.Size * 10
+	layerRootDir := filepath.Join(r.ClonePath(), layer.Digest.Hex())
+	layerDir := filepath.Join(path, layer.Digest.Hex())
+	err := os.MkdirAll(layerDir, 0700)
+	if err != nil {
+		return fmt.Errorf("could not create layer directory: %v", err)
+	}
+
+	var tarReader *tar.Reader
+
+	if strings.HasSuffix(strings.ToLower(layer.MediaType), "gzip") {
+		gzReader, err := gzip.NewReader(t)
+		if err != nil {
+			return fmt.Errorf("could not create gzip reader: %v", err)
+		}
+		tarReader = tar.NewReader(gzReader)
+		defer gzReader.Close()
+	} else if strings.HasSuffix(strings.ToLower(layer.MediaType), "zstd") {
+		zstdReader, err := zstd.NewReader(t)
+		if err != nil {
+			return fmt.Errorf("could not create zstd reader: %v", err)
+		}
+		tarReader = tar.NewReader(zstdReader)
+		defer zstdReader.Close()
+	} else {
+		tarReader = tar.NewReader(t)
+	}
+
+	for {
+		header, err := tarReader.Next()
+		if err == io.EOF {
+			break
+		} else if err != nil {
+			return fmt.Errorf("could not extract tar: %v", err)
+		}
+		path, err := fs.CleanJoin(layerRootDir, header.Name)
+		if err != nil {
+			logger.Error("%v - skipped", err)
+			continue
+		}
+		info := header.FileInfo()
+		if info.IsDir() {
+			if err = os.MkdirAll(path, 0700); err != nil {
+				return fmt.Errorf("could not create directory: %v", err)
+			}
+			continue
+		}
+		if info.Mode()&os.ModeSymlink != 0 {
+			logger.Warning("skipping file that is a symlink: %s", info.Name())
+			continue
+		}
+
+		file, err := os.OpenFile(path, os.O_CREATE|os.O_TRUNC|os.O_WRONLY, 0600) // #nosec G304
+		if err != nil {
+			logger.Error("%v", err)
+			continue
+		}
+		defer file.Close()
+
+		// Copy a maximum number of bytes (layer size * 10) so we do not get "bombs". It is unlikely that a file
+		// with significant entropy will be compressed more than 10x. We can review this.
+		n, err := io.CopyN(file, tarReader, size)
+		if err != nil && err != io.EOF {
+			return fmt.Errorf("could not copy file to disk: %v", err)
+		}
+		if n >= size {
+			logger.Warning("copying file %s did not finish due to max file size: %v", file.Name(), err)
+		}
```

Comment:
Not blocking the MR:

defer close would keep all those file handles built up in the loop until the function exits, correct?

If you wanted to close them right away but still use defer, you could make a `r.copyN(dest, src, n)` that just contains the content of this.